Assets/Mirror/Runtime/NetworkWriter.cs

```diff
@@ -26,120 +26,68 @@ public class NetworkWriter
-        /// </summary>
-        /// <param name="newLength"></param>
-        /// <remarks>
-        /// Zeros out any extra length created by setlength
-        /// </remarks>
-        public void SetLength(int newLength)
+        // create new writer with size
+        public NetworkWriter(int size)
         {
-            int oldLength = length;
-
-            // ensure length & capacity
-            EnsureLength(newLength);
-
-            // zero out new length
-            if (oldLength < newLength)
-            {
-                Array.Clear(buffer, oldLength, newLength - oldLength);
-            }
-
-            length = newLength;
-            position = Mathf.Min(position, length);
+            buffer = new byte[size];
         }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        void EnsureLength(int value)
-        {
-            if (length < value)
-            {
-                length = value;
-                EnsureCapacity(value);
-            }
-        }
-
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        void EnsureCapacity(int value)
-        {
-            if (buffer.Length < value)
-            {
-                int capacity = Math.Max(value, buffer.Length * 2);
-                Array.Resize(ref buffer, capacity);
-            }
-        }
-
-        // MemoryStream has 3 values: Position, Length and Capacity.
-        // Position is used to indicate where we are writing
-        // Length is how much data we have written
-        // capacity is how much memory we have allocated
-        // ToArray returns all the data we have written,  regardless of the current position
+        // ToArray copies all the data until .Position into an array.
         public byte[] ToArray()
         {
-            byte[] data = new byte[length];
-            Array.ConstrainedCopy(buffer, 0, data, 0, length);
+            byte[] data = new byte[Position];
+            Array.ConstrainedCopy(buffer, 0, data, 0, Position);
             return data;
         }
 
         // Gets the serialized data in an ArraySegment<byte>
         // this is similar to ToArray(),  but it gets the data in O(1)
         // and without allocations.
         // Do not write anything else or modify the NetworkWriter
-        // while you are using the ArraySegment
+        // while using the ArraySegment
         public ArraySegment<byte> ToArraySegment()
         {
-            return new ArraySegment<byte>(buffer, 0, length);
+            return new ArraySegment<byte>(buffer, 0, Position);
         }
 
         public void WriteByte(byte value)
         {
-            EnsureLength(position + 1);
-            buffer[position++] = value;
+            // enough space in buffer?
+            if (Space >= 1)
+            {
+                buffer[Position++] = value;
+            }
+            else throw new IndexOutOfRangeException($"Can't write 1 more byte to writer with Position={Position}, Capacity={Capacity}");
```

Comment:
Can we inverse this if,  eg ` if (not enough size) throw`.  Would make it easier to read than the if/else